htscodecs/arith_dynamic.c

```diff
@@ -843,20 +862,27 @@ unsigned char *arith_compress_to(unsigned char *in,  unsigned int in_size,
 //                        {1, 128}};
 
             for (j = 1; j <= m[MIN(i,3)][0]; j++) {
+                if (out2 - out > *out_size)
+                    continue; // an error, but caught in best_sz check later
+
                 olen2 = *out_size - (out2 - out);
                 //fprintf(stderr, "order=%d m=%d\n", order&3, m[MIN(i,4)][j]);
                 if ((order&3) == 0 && (m[MIN(i,3)][j]&1))
                     continue;
 
                 arith_compress_to(transposed+idx[i], part_len[i],
                                   out2, &olen2, m[MIN(i,3)][j] | X_NOSZ);
-                if (best_sz > olen2) {
+                if (olen2 && best_sz > olen2) {
```

Comment:
I'm not sure why you didn't check for a NULL return value here, as it would have needed fewer changes.  Anyway, checking `olen2` should now work here. However the `arith_compress_to()` recursion at line 888 assumes that it will work the second time round - I'm not sure if you could completely guarantee that as it could try to allocate memory for some of the options.  For safety it should probably be checked to ensure it worked.

Reply:
The reason for it is traditionally it's been acceptable for some compression methods to fail, eg attempting to stripe with an indivisible number of components.  (This may have been resolved now, but I'd have to check.)  The philosophy though is don't error if any call fails, but error if all fails.  It doesn't matter much how we determine failure (return value vs length), but I wanted to make length return 0 on failure anyway so they become the same thing really.  Perhaps a NULL check is easier to understand though (and safer incase of paths I missed or may miss in the future) so I'll consider changing it.

Reply:
Resolved with belt and braces both NULL and olen2.  It's not a critical part of the algorithm speed and it protects against future mistakes.